Compilation warnings

#3522 introduced some compilation warnings.

### Windows

```
2020-11-13T15:31:33.3469500Z -- Building for: Visual Studio 15 2017
2020-11-13T15:31:43.1306719Z -- The C compiler identification is MSVC 19.16.27043.0
2020-11-13T15:31:46.5396017Z -- The CXX compiler identification is MSVC 19.16.27043.0
2020-11-13T15:31:46.6305568Z -- Detecting C compiler ABI info
2020-11-13T15:31:50.0222750Z -- Detecting C compiler ABI info - done
2020-11-13T15:31:50.0286562Z -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe - skipped
2020-11-13T15:31:50.0292266Z -- Detecting C compile features
2020-11-13T15:31:50.0373341Z -- Detecting C compile features - done
2020-11-13T15:31:50.0488000Z -- Detecting CXX compiler ABI info
2020-11-13T15:31:53.1205054Z -- Detecting CXX compiler ABI info - done
2020-11-13T15:31:53.1211328Z -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/Hostx86/x64/cl.exe - skipped
2020-11-13T15:31:53.1216977Z -- Detecting CXX compile features
2020-11-13T15:31:53.1230295Z -- Detecting CXX compile features - done
2020-11-13T15:32:02.7442600Z -- Found OpenMP_C: -openmp (found version "2.0") 
2020-11-13T15:32:05.6884105Z -- Found OpenMP_CXX: -openmp (found version "2.0") 
2020-11-13T15:32:05.6885224Z -- Found OpenMP: TRUE (found version "2.0")  
2020-11-13T15:32:05.6894114Z -- Performing Test MM_PREFETCH
2020-11-13T15:32:08.4215399Z -- Performing Test MM_PREFETCH - Failed
2020-11-13T15:32:08.4216263Z -- Performing Test MM_MALLOC
2020-11-13T15:32:11.2883358Z -- Performing Test MM_MALLOC - Failed
2020-11-13T15:32:11.2937062Z -- Configuring done
2020-11-13T15:32:11.3352479Z -- Generating done
2020-11-13T15:32:11.3364843Z -- Build files have been written to: D:/a/1/s/build
2020-11-13T15:32:11.8442849Z Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
2020-11-13T15:32:11.8444089Z Copyright (C) Microsoft Corporation. All rights reserved.
2020-11-13T15:32:11.8444895Z 
2020-11-13T15:32:13.9216877Z   Checking Build System
2020-11-13T15:32:14.1538229Z   Building Custom Rule D:/a/1/s/CMakeLists.txt
2020-11-13T15:32:14.3317540Z   application.cpp
2020-11-13T15:32:14.3475734Z   boosting.cpp
2020-11-13T15:32:14.3485167Z   gbdt.cpp
2020-11-13T15:32:14.3486396Z   gbdt_model_text.cpp
2020-11-13T15:32:15.7928630Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\gbdt_model_text.cpp)D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\gbdt.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:15.7933677Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\gbdt_model_text.cpp)
2020-11-13T15:32:15.7935482Z   
2020-11-13T15:32:15.7940381Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\gbdt.cpp)
2020-11-13T15:32:15.7948555Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\boosting.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:15.8069430Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\boosting.cpp)
2020-11-13T15:32:15.8172196Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\application\application.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:15.8820495Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\application\application.cpp)
2020-11-13T15:32:22.3211522Z   gbdt_prediction.cpp
2020-11-13T15:32:23.4126949Z   prediction_early_stop.cpp
2020-11-13T15:32:24.4627697Z   bin.cpp
2020-11-13T15:32:24.5208242Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\gbdt_prediction.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:24.5831373Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\gbdt_prediction.cpp)
2020-11-13T15:32:24.8601229Z   config.cpp
2020-11-13T15:32:25.1709112Z   config_auto.cpp
2020-11-13T15:32:27.0988596Z   dataset.cpp
2020-11-13T15:32:30.7751573Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\dataset.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:31.1559888Z   dataset_loader.cpp
2020-11-13T15:32:31.3470876Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\dataset.cpp)
2020-11-13T15:32:33.7360961Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\dataset_loader.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:33.7481044Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\dataset_loader.cpp)
2020-11-13T15:32:34.3030834Z   file_io.cpp
2020-11-13T15:32:37.1994912Z   json11.cpp
2020-11-13T15:32:39.3002504Z   metadata.cpp
2020-11-13T15:32:42.1562994Z   parser.cpp
2020-11-13T15:32:42.8530031Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\metadata.cpp)train_share_states.cpp [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:42.8542081Z   
2020-11-13T15:32:43.0546955Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\metadata.cpp)
2020-11-13T15:32:43.5295534Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\parser.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:43.5305564Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\parser.cpp)
2020-11-13T15:32:43.6223301Z   tree.cpp
2020-11-13T15:32:43.8694890Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\train_share_states.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:43.9018411Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\train_share_states.cpp)
2020-11-13T15:32:43.9039444Z D:\a\1\s\src\io\train_share_states.cpp(248): warning C4458: declaration of 'is_col_wise' hides class member [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:44.0536319Z   D:\a\1\s\include\LightGBM/train_share_states.h(156): note: see declaration of 'LightGBM::TrainingShareStates::is_col_wise' (compiling source file D:\a\1\s\src\io\train_share_states.cpp)
2020-11-13T15:32:45.7615978Z   dcg_calculator.cpp
2020-11-13T15:32:46.0270908Z   metric.cpp
2020-11-13T15:32:51.5712638Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\tree.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:51.7026288Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\tree.cpp)
2020-11-13T15:32:51.7198029Z   ifaddrs_patch.cpp
2020-11-13T15:32:51.7207459Z   linker_topo.cpp
2020-11-13T15:32:51.8867895Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\metric\dcg_calculator.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:51.8870342Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\metric\dcg_calculator.cpp)
2020-11-13T15:32:51.8886852Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\metric\metric.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:51.8888968Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\metric\metric.cpp)
2020-11-13T15:32:51.8890508Z   linkers_mpi.cpp
2020-11-13T15:32:51.8891087Z   linkers_socket.cpp
2020-11-13T15:32:51.8891535Z   network.cpp
2020-11-13T15:32:51.8892630Z   objective_function.cpp
2020-11-13T15:32:52.1618239Z   cuda_tree_learner.cpp
2020-11-13T15:32:52.1761504Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\objective\objective_function.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:52.1763801Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\objective\objective_function.cpp)
2020-11-13T15:32:52.1914501Z   data_parallel_tree_learner.cpp
2020-11-13T15:32:52.9920030Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:52.9925592Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp)
2020-11-13T15:32:56.9824863Z   feature_parallel_tree_learner.cpp
2020-11-13T15:32:57.5979716Z   gpu_tree_learner.cpp
2020-11-13T15:32:57.6423936Z   serial_tree_learner.cpp
2020-11-13T15:32:57.9896958Z   tree_learner.cpp
2020-11-13T15:32:58.2450330Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\feature_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:58.4334799Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\feature_parallel_tree_learner.cpp)
2020-11-13T15:32:58.8144954Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\serial_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:58.8759650Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\serial_tree_learner.cpp)
2020-11-13T15:32:58.8768917Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:58.9705861Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\tree_learner.cpp)
2020-11-13T15:32:59.0107514Z D:\a\1\s\src\treelearner\serial_tree_learner.cpp(64): warning C4244: 'argument': conversion from 'uint64_t' to 'int', possible loss of data [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:32:59.2255366Z D:\a\1\s\src\treelearner\serial_tree_learner.cpp(137): warning C4244: 'argument': conversion from 'uint64_t' to 'int', possible loss of data [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:33:02.9344107Z   voting_parallel_tree_learner.cpp
2020-11-13T15:33:04.1215612Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:33:04.3145476Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp)
2020-11-13T15:33:05.1880418Z   c_api.cpp
2020-11-13T15:33:05.3194887Z D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp(71): warning C4244: 'initializing': conversion from 'uint64_t' to 'int', possible loss of data [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:33:05.3197002Z   D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp(22): note: while compiling class template member function 'void LightGBM::VotingParallelTreeLearner<LightGBM::CUDATreeLearner>::Init(const LightGBM::Dataset *,bool)'
2020-11-13T15:33:05.3202804Z   D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp(462): note: see reference to class template instantiation 'LightGBM::VotingParallelTreeLearner<LightGBM::CUDATreeLearner>' being compiled
2020-11-13T15:33:14.4288523Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\c_api.cpp) [D:\a\1\s\build\_lightgbm.vcxproj]
2020-11-13T15:33:14.4292280Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\c_api.cpp)
2020-11-13T15:33:29.8053211Z      Creating library D:/a/1/s/Release/lib_lightgbm.lib and object D:/a/1/s/Release/lib_lightgbm.exp
2020-11-13T15:33:30.4761116Z   _lightgbm.vcxproj -> D:\a\1\s\Release\lib_lightgbm.dll
2020-11-13T15:33:30.5696796Z   Building Custom Rule D:/a/1/s/CMakeLists.txt
2020-11-13T15:33:30.7267901Z   main.cpp
2020-11-13T15:33:30.7269724Z   application.cpp
2020-11-13T15:33:30.7303732Z   boosting.cpp
2020-11-13T15:33:30.7310346Z   gbdt.cpp
2020-11-13T15:33:31.7448760Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\boosting.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:31.7650176Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\boosting.cpp)
2020-11-13T15:33:31.7660295Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\gbdt.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:31.7729893Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\gbdt.cpp)
2020-11-13T15:33:31.7737308Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\application\application.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:31.9012423Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\application\application.cpp)
2020-11-13T15:33:34.7457131Z   gbdt_model_text.cpp
2020-11-13T15:33:35.7498007Z   gbdt_prediction.cpp
2020-11-13T15:33:37.9639115Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\gbdt_prediction.cpp)prediction_early_stop.cpp [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:37.9649382Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\boosting\gbdt_model_text.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:38.0468193Z   
2020-11-13T15:33:38.2057733Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\gbdt_model_text.cpp)
2020-11-13T15:33:38.3828679Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\boosting\gbdt_prediction.cpp)
2020-11-13T15:33:38.4312241Z   bin.cpp
2020-11-13T15:33:38.8641966Z   config.cpp
2020-11-13T15:33:39.6772167Z   config_auto.cpp
2020-11-13T15:33:44.1373010Z   dataset.cpp
2020-11-13T15:33:46.4274466Z   dataset_loader.cpp
2020-11-13T15:33:47.3409324Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\dataset.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:48.1235205Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\dataset.cpp)
2020-11-13T15:33:48.3138092Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\dataset_loader.cpp)file_io.cpp [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:48.5050003Z   
2020-11-13T15:33:54.0406900Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\dataset_loader.cpp)
2020-11-13T15:33:54.0408207Z   json11.cpp
2020-11-13T15:33:54.0408641Z   metadata.cpp
2020-11-13T15:33:54.0409743Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\metadata.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:54.0411679Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\metadata.cpp)
2020-11-13T15:33:54.0412726Z   parser.cpp
2020-11-13T15:33:55.9313973Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\parser.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:56.1434414Z   train_share_states.cpp
2020-11-13T15:33:56.4063793Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\parser.cpp)
2020-11-13T15:33:56.9396996Z   tree.cpp
2020-11-13T15:33:56.9398119Z   dcg_calculator.cpp
2020-11-13T15:33:57.3377709Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\train_share_states.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:57.3389446Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\train_share_states.cpp)
2020-11-13T15:33:57.4621325Z D:\a\1\s\src\io\train_share_states.cpp(248): warning C4458: declaration of 'is_col_wise' hides class member [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:57.5483046Z   D:\a\1\s\include\LightGBM/train_share_states.h(156): note: see declaration of 'LightGBM::TrainingShareStates::is_col_wise' (compiling source file D:\a\1\s\src\io\train_share_states.cpp)
2020-11-13T15:33:57.7143247Z   metric.cpp
2020-11-13T15:33:57.7222865Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\io\tree.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:57.8423915Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\io\tree.cpp)
2020-11-13T15:33:57.8960630Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\metric\dcg_calculator.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:57.9293996Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\metric\dcg_calculator.cpp)
2020-11-13T15:33:58.4835193Z   ifaddrs_patch.cpp
2020-11-13T15:33:58.4953043Z   linker_topo.cpp
2020-11-13T15:33:58.6081715Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\metric\metric.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:33:58.6331345Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\metric\metric.cpp)
2020-11-13T15:33:59.6150568Z   linkers_mpi.cpp
2020-11-13T15:33:59.6576789Z   linkers_socket.cpp
2020-11-13T15:34:02.2136520Z   network.cpp
2020-11-13T15:34:03.7045200Z   objective_function.cpp
2020-11-13T15:34:03.8148869Z   cuda_tree_learner.cpp
2020-11-13T15:34:03.9840891Z   data_parallel_tree_learner.cpp
2020-11-13T15:34:04.6361617Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\objective\objective_function.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:04.6375597Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\objective\objective_function.cpp)
2020-11-13T15:34:04.9378749Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:04.9381805Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\data_parallel_tree_learner.cpp)
2020-11-13T15:34:06.5237413Z   feature_parallel_tree_learner.cpp
2020-11-13T15:34:06.9096639Z   gpu_tree_learner.cpp
2020-11-13T15:34:06.9436900Z   serial_tree_learner.cpp
2020-11-13T15:34:08.4423646Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\feature_parallel_tree_learner.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:08.9167174Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\feature_parallel_tree_learner.cpp)
2020-11-13T15:34:10.6793081Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\serial_tree_learner.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:10.7216672Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\serial_tree_learner.cpp)
2020-11-13T15:34:10.9052450Z   tree_learner.cpp
2020-11-13T15:34:10.9081768Z D:\a\1\s\src\treelearner\serial_tree_learner.cpp(64): warning C4244: 'argument': conversion from 'uint64_t' to 'int', possible loss of data [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:11.0678000Z D:\a\1\s\src\treelearner\serial_tree_learner.cpp(137): warning C4244: 'argument': conversion from 'uint64_t' to 'int', possible loss of data [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:12.0850613Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\tree_learner.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:12.2493072Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\tree_learner.cpp)
2020-11-13T15:34:16.3848253Z   voting_parallel_tree_learner.cpp
2020-11-13T15:34:17.9546035Z D:\a\1\s\include\LightGBM/train_share_states.h(154): warning C4099: 'LightGBM::TrainingShareStates': type name first seen using 'class' now seen using 'struct' (compiling source file D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp) [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:17.9555344Z   D:\a\1\s\include\LightGBM/train_share_states.h(154): note: see declaration of 'LightGBM::TrainingShareStates' (compiling source file D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp)
2020-11-13T15:34:20.8637978Z D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp(71): warning C4244: 'initializing': conversion from 'uint64_t' to 'int', possible loss of data [D:\a\1\s\build\lightgbm.vcxproj]
2020-11-13T15:34:21.0799490Z   D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp(22): note: while compiling class template member function 'void LightGBM::VotingParallelTreeLearner<LightGBM::CUDATreeLearner>::Init(const LightGBM::Dataset *,bool)'
2020-11-13T15:34:21.4555644Z   D:\a\1\s\src\treelearner\voting_parallel_tree_learner.cpp(462): note: see reference to class template instantiation 'LightGBM::VotingParallelTreeLearner<LightGBM::CUDATreeLearner>' being compiled
2020-11-13T15:34:38.7699275Z      Creating library D:/a/1/s/Release/lightgbm.lib and object D:/a/1/s/Release/lightgbm.exp
2020-11-13T15:34:39.0920882Z   lightgbm.vcxproj -> D:\a\1\s\Release\lightgbm.exe
2020-11-13T15:34:39.1414143Z   Building Custom Rule D:/a/1/s/CMakeLists.txt
```


### Clang

```
2020-11-13T15:29:15.4731240Z -- The C compiler identification is AppleClang 9.1.0.9020039
2020-11-13T15:29:15.5506550Z -- The CXX compiler identification is AppleClang 9.1.0.9020039
2020-11-13T15:29:15.5869810Z -- Detecting C compiler ABI info
2020-11-13T15:29:16.3579670Z -- Detecting C compiler ABI info - done
2020-11-13T15:29:16.3843050Z -- Check for working C compiler: /Applications/Xcode_9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
2020-11-13T15:29:16.3845950Z -- Detecting C compile features
2020-11-13T15:29:16.3918560Z -- Detecting C compile features - done
2020-11-13T15:29:16.4037050Z -- Detecting CXX compiler ABI info
2020-11-13T15:29:16.5791430Z -- Detecting CXX compiler ABI info - done
2020-11-13T15:29:16.5929250Z -- Check for working CXX compiler: /Applications/Xcode_9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
2020-11-13T15:29:16.6029190Z -- Detecting CXX compile features
2020-11-13T15:29:16.6088540Z -- Detecting CXX compile features - done
2020-11-13T15:29:18.3132310Z -- Found OpenMP_C: -Xclang -fopenmp (found version "3.1") 
2020-11-13T15:29:18.5109710Z -- Found OpenMP_CXX: -Xclang -fopenmp (found version "3.1") 
2020-11-13T15:29:18.5111230Z -- Found OpenMP: TRUE (found version "3.1")  
2020-11-13T15:29:18.5134330Z -- Performing Test MM_PREFETCH
2020-11-13T15:29:18.7267190Z -- Performing Test MM_PREFETCH - Success
2020-11-13T15:29:18.7268240Z -- Using _mm_prefetch
2020-11-13T15:29:18.7273530Z -- Performing Test MM_MALLOC
2020-11-13T15:29:18.8964770Z -- Performing Test MM_MALLOC - Success
2020-11-13T15:29:18.8965910Z -- Using _mm_malloc
2020-11-13T15:29:18.8987610Z -- Configuring done
2020-11-13T15:29:18.9273190Z -- Generating done
2020-11-13T15:29:18.9302960Z -- Build files have been written to: /Users/runner/work/1/s/build
2020-11-13T15:29:19.0407440Z Scanning dependencies of target _lightgbm
2020-11-13T15:29:19.0636760Z [  2%] Building CXX object CMakeFiles/_lightgbm.dir/src/application/application.cpp.o
2020-11-13T15:29:19.0738720Z [  5%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt.cpp.o
2020-11-13T15:29:19.0859000Z [  8%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/boosting.cpp.o
2020-11-13T15:29:19.0927910Z [ 11%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt_model_text.cpp.o
2020-11-13T15:29:20.4095060Z In file included from /Users/runner/work/1/s/src/application/application.cpp:8:
2020-11-13T15:29:20.4195450Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:20.4204820Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:20.4206170Z struct TrainingShareStates {
2020-11-13T15:29:20.4207070Z ^
2020-11-13T15:29:20.4208070Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:20.4209050Z class TrainingShareStates;
2020-11-13T15:29:20.4210510Z ^~~~~
2020-11-13T15:29:20.4211270Z struct
2020-11-13T15:29:20.4212230Z In file included from /Users/runner/work/1/s/src/boosting/gbdt_model_text.cpp:6:
2020-11-13T15:29:20.4213330Z In file included from /Users/runner/work/1/s/include/LightGBM/metric.h:9:
2020-11-13T15:29:20.4275230Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:20.4377150Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:20.4476920Z struct TrainingShareStates {
2020-11-13T15:29:20.4531280Z ^
2020-11-13T15:29:20.4564700Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:20.4670740Z class TrainingShareStates;
2020-11-13T15:29:20.4772610Z ^~~~~
2020-11-13T15:29:21.3761940Z struct
2020-11-13T15:29:21.3762690Z In file included from /Users/runner/work/1/s/src/boosting/gbdt.cpp:5:
2020-11-13T15:29:21.3863920Z In file included from /Users/runner/work/1/s/src/boosting/gbdt.h:9:
2020-11-13T15:29:21.3918540Z In file included from /Users/runner/work/1/s/include/LightGBM/objective_function.h:9:
2020-11-13T15:29:21.3931490Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:21.3946810Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:21.4048180Z struct TrainingShareStates {
2020-11-13T15:29:21.4149780Z ^
2020-11-13T15:29:21.4251650Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:21.4265630Z class TrainingShareStates;
2020-11-13T15:29:21.4332680Z ^~~~~
2020-11-13T15:29:21.4366590Z struct
2020-11-13T15:29:21.4466340Z In file included from /Users/runner/work/1/s/src/boosting/boosting.cpp:7:
2020-11-13T15:29:21.4471210Z In file included from /Users/runner/work/1/s/src/boosting/dart.hpp:16:
2020-11-13T15:29:21.4501900Z In file included from /Users/runner/work/1/s/src/boosting/gbdt.h:9:
2020-11-13T15:29:21.4526200Z In file included from /Users/runner/work/1/s/include/LightGBM/objective_function.h:9:
2020-11-13T15:29:21.4557330Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:21.4602170Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:21.4703830Z struct TrainingShareStates {
2020-11-13T15:29:21.4779740Z ^
2020-11-13T15:29:21.4790090Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:21.4802240Z class TrainingShareStates;
2020-11-13T15:29:21.4802760Z ^~~~~
2020-11-13T15:29:21.4805150Z struct
2020-11-13T15:29:24.0839140Z 1 warning generated.
2020-11-13T15:29:24.1045380Z [ 14%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/gbdt_prediction.cpp.o
2020-11-13T15:29:25.1580250Z 1 warning generated.
2020-11-13T15:29:25.1821140Z [ 17%] Building CXX object CMakeFiles/_lightgbm.dir/src/boosting/prediction_early_stop.cpp.o
2020-11-13T15:29:25.3857460Z In file included from /Users/runner/work/1/s/src/boosting/gbdt_prediction.cpp:5:
2020-11-13T15:29:25.3959750Z In file included from /Users/runner/work/1/s/include/LightGBM/objective_function.h:9:
2020-11-13T15:29:25.3978310Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:25.4062470Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:25.4079460Z struct TrainingShareStates {
2020-11-13T15:29:25.4205360Z ^
2020-11-13T15:29:25.4307200Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:25.4404520Z class TrainingShareStates;
2020-11-13T15:29:25.4489010Z ^~~~~
2020-11-13T15:29:25.4577950Z struct
2020-11-13T15:29:25.7273360Z 1 warning generated.
2020-11-13T15:29:25.7485070Z [ 20%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/bin.cpp.o
2020-11-13T15:29:26.4555530Z 1 warning generated.
2020-11-13T15:29:26.4756330Z [ 23%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/config.cpp.o
2020-11-13T15:29:26.6624960Z [ 26%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/config_auto.cpp.o
2020-11-13T15:29:28.0210600Z 1 warning generated.
2020-11-13T15:29:28.0505310Z [ 29%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/dataset.cpp.o
2020-11-13T15:29:29.2413240Z In file included from /Users/runner/work/1/s/src/io/dataset.cpp:6:
2020-11-13T15:29:29.2488350Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:29.2490820Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:29.2492120Z struct TrainingShareStates {
2020-11-13T15:29:29.2492880Z ^
2020-11-13T15:29:29.2493800Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:29.2494720Z class TrainingShareStates;
2020-11-13T15:29:29.2495550Z ^~~~~
2020-11-13T15:29:29.2496310Z struct
2020-11-13T15:29:31.3908820Z [ 32%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/dataset_loader.cpp.o
2020-11-13T15:29:31.7217190Z [ 35%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/file_io.cpp.o
2020-11-13T15:29:32.2340140Z In file included from /Users/runner/work/1/s/src/io/dataset_loader.cpp:5:
2020-11-13T15:29:32.2341270Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset_loader.h:8:
2020-11-13T15:29:32.2441290Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:32.2443630Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:32.2480020Z struct TrainingShareStates {
2020-11-13T15:29:32.2580600Z ^
2020-11-13T15:29:32.2581550Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:32.2682060Z class TrainingShareStates;
2020-11-13T15:29:32.2740200Z ^~~~~
2020-11-13T15:29:32.2840630Z struct
2020-11-13T15:29:32.3347500Z [ 38%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/json11.cpp.o
2020-11-13T15:29:34.4478810Z [ 41%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/metadata.cpp.o
2020-11-13T15:29:34.5091360Z 1 warning generated.
2020-11-13T15:29:34.5307810Z [ 44%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/parser.cpp.o
2020-11-13T15:29:35.2995850Z In file included from /Users/runner/work/1/s/src/io/metadata.cpp:5:
2020-11-13T15:29:35.3096230Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:35.3198940Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:35.3288910Z struct TrainingShareStates {
2020-11-13T15:29:35.3290350Z ^
2020-11-13T15:29:35.3291370Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:35.3292240Z class TrainingShareStates;
2020-11-13T15:29:35.3292860Z ^~~~~
2020-11-13T15:29:35.3393770Z struct
2020-11-13T15:29:35.3997980Z In file included from /Users/runner/work/1/s/src/io/parser.cpp:5:
2020-11-13T15:29:35.4098360Z In file included from /Users/runner/work/1/s/src/io/parser.hpp:8:
2020-11-13T15:29:35.4200140Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:35.4207260Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:35.4307770Z struct TrainingShareStates {
2020-11-13T15:29:35.4376630Z ^
2020-11-13T15:29:35.4462820Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:35.4564320Z class TrainingShareStates;
2020-11-13T15:29:35.4589180Z ^~~~~
2020-11-13T15:29:35.4692310Z struct
2020-11-13T15:29:37.1012750Z 1 warning generated.
2020-11-13T15:29:37.1205420Z [ 47%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/train_share_states.cpp.o
2020-11-13T15:29:38.0137860Z 1 warning generated.
2020-11-13T15:29:38.0339820Z [ 50%] Building CXX object CMakeFiles/_lightgbm.dir/src/io/tree.cpp.o
2020-11-13T15:29:38.2534080Z In file included from /Users/runner/work/1/s/src/io/train_share_states.cpp:7:
2020-11-13T15:29:38.2595930Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:38.2680760Z struct TrainingShareStates {
2020-11-13T15:29:38.2773660Z ^
2020-11-13T15:29:38.2870890Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:38.2937720Z class TrainingShareStates;
2020-11-13T15:29:38.3086250Z ^~~~~
2020-11-13T15:29:38.3115470Z struct
2020-11-13T15:29:38.5050720Z In file included from /Users/runner/work/1/s/src/io/train_share_states.cpp:7:
2020-11-13T15:29:38.5052580Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:134:16: warning: private field 'cur_multi_val_bin_' is not used [-Wunused-private-field]
2020-11-13T15:29:38.5053820Z   MultiValBin* cur_multi_val_bin_;
2020-11-13T15:29:38.5054740Z                ^
2020-11-13T15:29:38.5363210Z [ 52%] Building CXX object CMakeFiles/_lightgbm.dir/src/metric/dcg_calculator.cpp.o
2020-11-13T15:29:38.9322170Z 2 warnings generated.
2020-11-13T15:29:38.9517580Z [ 55%] Building CXX object CMakeFiles/_lightgbm.dir/src/metric/metric.cpp.o
2020-11-13T15:29:38.9941810Z In file included from /Users/runner/work/1/s/src/io/tree.cpp:5:
2020-11-13T15:29:38.9942690Z In file included from /Users/runner/work/1/s/include/LightGBM/tree.h:8:
2020-11-13T15:29:38.9943240Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:38.9944870Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:38.9945770Z struct TrainingShareStates {
2020-11-13T15:29:38.9946100Z ^
2020-11-13T15:29:38.9986370Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:38.9987110Z class TrainingShareStates;
2020-11-13T15:29:38.9987440Z ^~~~~
2020-11-13T15:29:38.9987910Z struct
2020-11-13T15:29:39.2829040Z 1 warning generated.
2020-11-13T15:29:39.3120550Z [ 58%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/ifaddrs_patch.cpp.o
2020-11-13T15:29:39.3456950Z [ 61%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linker_topo.cpp.o
2020-11-13T15:29:39.4905270Z In file included from /Users/runner/work/1/s/src/metric/dcg_calculator.cpp:5:
2020-11-13T15:29:39.4992560Z In file included from /Users/runner/work/1/s/include/LightGBM/metric.h:9:
2020-11-13T15:29:39.4994080Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:39.4996490Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:39.4997710Z struct TrainingShareStates {
2020-11-13T15:29:39.4998510Z ^
2020-11-13T15:29:39.4999420Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:39.5000380Z class TrainingShareStates;
2020-11-13T15:29:39.5001140Z ^~~~~
2020-11-13T15:29:39.5102400Z struct
2020-11-13T15:29:40.1754880Z In file included from /Users/runner/work/1/s/src/metric/metric.cpp:5:
2020-11-13T15:29:40.1857580Z In file included from /Users/runner/work/1/s/include/LightGBM/metric.h:9:
2020-11-13T15:29:40.1874060Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:40.1979330Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:40.2013390Z struct TrainingShareStates {
2020-11-13T15:29:40.2030280Z ^
2020-11-13T15:29:40.2131600Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:40.2170330Z class TrainingShareStates;
2020-11-13T15:29:40.2249650Z ^~~~~
2020-11-13T15:29:40.2276450Z struct
2020-11-13T15:29:41.0731840Z [ 64%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linkers_mpi.cpp.o
2020-11-13T15:29:41.1028420Z [ 67%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/linkers_socket.cpp.o
2020-11-13T15:29:41.4348100Z 1 warning generated.
2020-11-13T15:29:41.4556920Z [ 70%] Building CXX object CMakeFiles/_lightgbm.dir/src/network/network.cpp.o
2020-11-13T15:29:43.7266690Z [ 73%] Building CXX object CMakeFiles/_lightgbm.dir/src/objective/objective_function.cpp.o
2020-11-13T15:29:44.9592560Z In file included from /Users/runner/work/1/s/src/objective/objective_function.cpp:5:
2020-11-13T15:29:44.9595670Z In file included from /Users/runner/work/1/s/include/LightGBM/objective_function.h:9:
2020-11-13T15:29:44.9597290Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:44.9673860Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:44.9705380Z struct TrainingShareStates {
2020-11-13T15:29:44.9776030Z ^
2020-11-13T15:29:44.9842490Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:44.9866140Z class TrainingShareStates;
2020-11-13T15:29:44.9867070Z ^~~~~
2020-11-13T15:29:44.9867900Z struct
2020-11-13T15:29:45.0195160Z [ 76%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/cuda_tree_learner.cpp.o
2020-11-13T15:29:45.0510670Z [ 79%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/data_parallel_tree_learner.cpp.o
2020-11-13T15:29:45.1059980Z 1 warning generated.
2020-11-13T15:29:45.1289570Z [ 82%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/feature_parallel_tree_learner.cpp.o
2020-11-13T15:29:45.9352780Z In file included from /Users/runner/work/1/s/src/treelearner/data_parallel_tree_learner.cpp:9:
2020-11-13T15:29:45.9454100Z In file included from /Users/runner/work/1/s/src/treelearner/parallel_tree_learner.h:15:
2020-11-13T15:29:45.9554700Z In file included from /Users/runner/work/1/s/src/treelearner/cuda_tree_learner.h:10:
2020-11-13T15:29:45.9606950Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:45.9709630Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:45.9812010Z struct TrainingShareStates {
2020-11-13T15:29:45.9913340Z ^
2020-11-13T15:29:46.0015030Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:46.0086880Z class TrainingShareStates;
2020-11-13T15:29:46.0087670Z ^~~~~
2020-11-13T15:29:46.0113780Z struct
2020-11-13T15:29:46.0322480Z In file included from /Users/runner/work/1/s/src/treelearner/feature_parallel_tree_learner.cpp:8:
2020-11-13T15:29:46.0347940Z In file included from /Users/runner/work/1/s/src/treelearner/parallel_tree_learner.h:15:
2020-11-13T15:29:46.0448510Z In file included from /Users/runner/work/1/s/src/treelearner/cuda_tree_learner.h:10:
2020-11-13T15:29:46.0472660Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:46.0516140Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:46.0547450Z struct TrainingShareStates {
2020-11-13T15:29:46.0626780Z ^
2020-11-13T15:29:46.0693010Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:46.0762580Z class TrainingShareStates;
2020-11-13T15:29:46.0763890Z ^~~~~
2020-11-13T15:29:46.0764690Z struct
2020-11-13T15:29:46.2469590Z 1 warning generated.
2020-11-13T15:29:46.2680290Z [ 85%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/gpu_tree_learner.cpp.o
2020-11-13T15:29:46.2995540Z [ 88%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/serial_tree_learner.cpp.o
2020-11-13T15:29:47.3751430Z In file included from /Users/runner/work/1/s/src/treelearner/serial_tree_learner.cpp:5:
2020-11-13T15:29:47.3852390Z In file included from /Users/runner/work/1/s/src/treelearner/serial_tree_learner.h:8:
2020-11-13T15:29:47.3853290Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:47.3878660Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:47.3907010Z struct TrainingShareStates {
2020-11-13T15:29:47.3955260Z ^
2020-11-13T15:29:47.3956530Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:47.3957550Z class TrainingShareStates;
2020-11-13T15:29:47.3958370Z ^~~~~
2020-11-13T15:29:47.3959140Z struct
2020-11-13T15:29:52.2724540Z 1 warning generated.
2020-11-13T15:29:52.2946520Z [ 91%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/tree_learner.cpp.o
2020-11-13T15:29:53.5634470Z In file included from /Users/runner/work/1/s/src/treelearner/tree_learner.cpp:7:
2020-11-13T15:29:53.5635900Z In file included from /Users/runner/work/1/s/src/treelearner/cuda_tree_learner.h:10:
2020-11-13T15:29:53.5636580Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:29:53.5638210Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:29:53.5639200Z struct TrainingShareStates {
2020-11-13T15:29:53.5639810Z ^
2020-11-13T15:29:53.5640510Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:29:53.5641250Z class TrainingShareStates;
2020-11-13T15:29:53.5641830Z ^~~~~
2020-11-13T15:29:53.5642340Z struct
2020-11-13T15:30:03.4449310Z 1 warning generated.
2020-11-13T15:30:03.4907550Z [ 94%] Building CXX object CMakeFiles/_lightgbm.dir/src/treelearner/voting_parallel_tree_learner.cpp.o
2020-11-13T15:30:04.3856160Z In file included from /Users/runner/work/1/s/src/treelearner/voting_parallel_tree_learner.cpp:11:
2020-11-13T15:30:04.3957080Z In file included from /Users/runner/work/1/s/src/treelearner/parallel_tree_learner.h:15:
2020-11-13T15:30:04.4058090Z In file included from /Users/runner/work/1/s/src/treelearner/cuda_tree_learner.h:10:
2020-11-13T15:30:04.4064020Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:30:04.4114470Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:30:04.4173570Z struct TrainingShareStates {
2020-11-13T15:30:04.4204160Z ^
2020-11-13T15:30:04.4207440Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:30:04.4254600Z class TrainingShareStates;
2020-11-13T15:30:04.4264570Z ^~~~~
2020-11-13T15:30:04.4299320Z struct
2020-11-13T15:30:04.7244320Z 1 warning generated.
2020-11-13T15:30:04.7725380Z [ 97%] Building CXX object CMakeFiles/_lightgbm.dir/src/c_api.cpp.o
2020-11-13T15:30:05.8163910Z In file included from /Users/runner/work/1/s/src/c_api.cpp:9:
2020-11-13T15:30:05.8200540Z In file included from /Users/runner/work/1/s/include/LightGBM/dataset.h:11:
2020-11-13T15:30:05.8210940Z /Users/runner/work/1/s/include/LightGBM/train_share_states.h:154:1: warning: 'TrainingShareStates' defined as a struct here but previously declared as a class [-Wmismatched-tags]
2020-11-13T15:30:05.8265890Z struct TrainingShareStates {
2020-11-13T15:30:05.8318280Z ^
2020-11-13T15:30:05.8419540Z /Users/runner/work/1/s/include/LightGBM/feature_group.h:21:1: note: did you mean struct here?
2020-11-13T15:30:05.8422500Z class TrainingShareStates;
2020-11-13T15:30:05.8423350Z ^~~~~
2020-11-13T15:30:05.8424140Z struct
2020-11-13T15:30:09.4335810Z 1 warning generated.
2020-11-13T15:30:15.8063850Z 1 warning generated.
2020-11-13T15:30:25.7220380Z 1 warning generated.
2020-11-13T15:30:34.0996810Z 1 warning generated.
2020-11-13T15:30:34.1485060Z [100%] Linking CXX shared library ../lib_lightgbm.so
2020-11-13T15:30:34.3151540Z [100%] Built target _lightgbm
```

--- a/include/LightGBM/feature_group.h
+++ b/include/LightGBM/feature_group.h
@@ -18,7 +18,7 @@ namespace LightGBM {
 
 class Dataset;
 class DatasetLoader;
-class TrainingShareStates;
+struct TrainingShareStates;
 class MultiValBinWrapper;
 /*! \brief Using to store data and providing some operations on one feature
  * group*/

